gpytorch/lazy/cat_lazy_tensor.py

```diff
@@ -0,0 +1,312 @@
+#!/usr/bin/env python3
+
+import torch
+import itertools
+from collections import defaultdict
+from .lazy_tensor import LazyTensor
+from .non_lazy_tensor import NonLazyTensor
+
+
+class CatLazyTensor(LazyTensor):
+    r"""
+    A `LazyTensor` that represents the concatenation of other lazy tensors.
+    Each LazyTensor must have the same shape except in the concatenating
+    dimension.
+
+    Args:
+        - :attr:`lazy_tensors` (list of LazyTensors):
+            A list of LazyTensors whose sizes are the same except in
+            concatenating dimension :attr:`dim`
+        - :attr:`dim` (int):
+            The concatenating dimension which can be a batch dimension.
+        - :attr:`output_device` (torch.device):
+            The CatLazyTensor will appear to appear on :attr:`output_device`
+            and place any output `torch.Tensors` on :attr:`output_device`
+    """
+
+    def __init__(self, *lazy_tensors, dim=0, output_device=None):
+        if len(lazy_tensors) == 0:
+            raise RuntimeError("List of LazyTensors must be non-empty")
+        if not all([isinstance(t, LazyTensor) for t in lazy_tensors]):
+            raise RuntimeError("CatLazyTensor requires a list of all LazyTensors")
+
+        if len(lazy_tensors) == 1:
+            dim = 0
```

Comment:
We need this in order to handle certain base cases correctly when we're slicing a CatLazyTensor

Reply:
I removed that check, made the corresponding section of `_getitem` just return the single LazyTensor instead of forming a `CatLazyTensor` out of it, and explicitly included a `RuntimeError` if someone tries to instantiate a `CatLazyTensor` with only one tensor: you should not ever do that.